### PR TITLE
feat: added transitive dependencies

### PR DESCRIPTION
**Pull request by MichaelYochpaz**
In order to extend our vulnerability scanning feature to support child dependencies for [npm packages](https://docs.npmjs.com/cli/v6/configuring-npm/package-json) as described in #5 , this pull request updates the package service to return all nested dependencies on the internal `/package` endpoint for consumption by vulnerability service instead of just returning the versions for the first level.

It supports a `GET` request to the `/package/:name/:version` endpoint and will return a JSON structure representing the full tree of dependencies. We will always return the latest matching version of a package supported to mimic the behavior of a fresh `npm install`.

**Testing**

It can be tested locally by making a request for a package with sub-dependencies e.g. `react@16.13.0`

```
curl -s http://localhost:3000/package/react/16.13.0 | jq .
```

**Related Ticket**

    * [Support returning the full dependency tree for packages #5](https://github.com/snyk/snyk-code-review-exercise/issues/5)

### CODE REVIEW

File: packages/modules/npm.py
Comment (MichaelYochpaz) [inline packages/modules/npm.py:1]: Please add docstrings to both functions

File: packages/modules/npm.py [hunk 1]
```diff
@@ -1,21 +1,44 @@
 import requests
+import semver
 
 from packages.models import VersionedPackage
 
 NPM_REGISTRY_URL = "https://registry.npmjs.org"
```
Comment (MichaelYochpaz) [inline packages/modules/npm.py:6]: Should we perhaps move this to the settings.py module?
(This is a nitpick)

File: packages/modules/npm.py [hunk 2]
```diff
@@ -1,21 +1,44 @@
...
 
 
-def get_versioned_package(name: str, version: str) -> VersionedPackage:
-    url = f"{NPM_REGISTRY_URL}/{name}/{version}"
+def get_package(name: str, range: str) -> VersionedPackage:
```
Comment (MichaelYochpaz) [inline packages/modules/npm.py:9]: The ticket (correctly) mentioned caching. Let's add a `@lru_cache(maxsize=1000)` decorator to implement it.
This will return the same result from cache for any combination of similar name and range that was cached.

The default `maxsize` is 128, and I believe we could benefit from a higher size for better performance on packages with larger dependency trees. Memory shouldn't be an issue with that size, but please check and validate that.

File: packages/modules/npm.py [hunk 3]
```diff
@@ -1,21 +1,44 @@
...
+    url = f"{NPM_REGISTRY_URL}/{name}"
 
-    response = requests.get(url)
-    npm_package = response.json()
+    npm_package = requests.get(url).json()
+    versions = list(npm_package["versions"].keys())
+    version = semver.min_satisfying(versions, range)
+    version_record = npm_package["versions"][version]
```
Comment (MichaelYochpaz) [inline packages/modules/npm.py:15]: No error handling here. Please update the code and check for error responses (for example, if an invalid package name is used, which currently causes crashes, as the ticket described).

Also handle cases where the responses (errors) aren't JSON, and add appropriate transparent error messages to the end user (different error messages if there's a timeout, and if an invalid name was used).

File: packages/modules/npm.py [hunk 4]
```diff
@@ -1,21 +1,44 @@
...
+
+
+def request_package(name: str, range: str) -> tuple[VersionedPackage, dict]:
+    url = f"{NPM_REGISTRY_URL}/{name}"
+
+    npm_package = requests.get(url).json()
+
+    versions = list(npm_package["versions"].keys())
+    version = semver.min_satisfying(versions, range)
+    version_record = npm_package["versions"][version]
```
Comment (MichaelYochpaz) [inline packages/modules/npm.py:38]: Same issue here. Please add error handling here as well.

File: packages/modules/npm.py
Comment (MichaelYochpaz) [inline packages/modules/npm.py:1]: Please add docstrings to both functions

File: packages/modules/npm.py (hunk 2 above)
Comment (MichaelYochpaz) [inline packages/modules/npm.py:9]: Let's add caching (as the ticket required). I've set the max caching size to 1,000 (over the default 128), I think that should give a decent performance boost for large dependency trees, and shouldn't use too much memory. Please check.

```suggestion
@lru_cache(maxsize=1000)
def get_package(name: str, range: str) -> VersionedPackage:
```

File: packages/modules/npm.py (hunk 3 above)
Comment (MichaelYochpaz) [inline packages/modules/npm.py:15]: There's no error-handling. That's probably what causes the crashing issues mentioned on the ticket.
Please handle errors (where the response might not be in JSON format).

```suggestion
    npm_package_response = requests.get(url)
    
    try:
        npm_package_response.raise_for_status()
        npm_package = npm_package_response.json()
    except Exception:  # Preferable update this with specific excppetions
        # Add logic here
    versions = list(npm_package["versions"].keys())
    version = semver.min_satisfying(versions, range)
    version_record = npm_package["versions"][version]
```

File: packages/modules/npm.py [hunk 5]
```diff
@@ -1,21 +1,44 @@
...
 
 
-def get_versioned_package(name: str, version: str) -> VersionedPackage:
-    url = f"{NPM_REGISTRY_URL}/{name}/{version}"
+def get_package(name: str, range: str) -> VersionedPackage:
+    url = f"{NPM_REGISTRY_URL}/{name}"
 
-    response = requests.get(url)
-    npm_package = response.json()
+    npm_package = requests.get(url).json()
```
Comment (MichaelYochpaz) [inline packages/modules/npm.py:12]: There's no error-handling. That's probably what causes the crashing issues mentioned on the ticket.
Please handle errors (where the response might not be in JSON format).

```suggestion
    npm_package_response = requests.get(url)
    
    try:
        npm_package_response.raise_for_status()
        npm_package = npm_package_response.json()
    except Exception:  # Preferable update this with specific excppetions
        # Add logic here
```

File: packages/modules/npm.py
Comment (MichaelYochpaz) [inline packages/modules/npm.py:1]: This will be much more performant and run faster by big margins on large dependency treed if we use async to fetch dependencies data.

Let's rewrite it in async. Note that if the `get_package` function turns `async`, a different caching method should be used, as `lru_cache` doesn't work properly with async functions (do some research).

**Comment by MichaelYochpaz** [comment]
Async code sample:

```python
import asyncio
import semver
from functools import lru_cache

import httpx

from packages.models import VersionedPackage

NPM_REGISTRY_URL = "https://registry.npmjs.org"

# NOTE: Make sure to gracefully close the client after using it

async def get_package(client: httpx.AsyncClient, name: str, range: str) -> VersionedPackage:
    package, dependencies_dict = await request_package(client=client, name=name, range=range)
    
    if dependencies_dict:
        package.dependencies = await fetch_all_dependencies(dependencies=dependencies_dict)
    
    return package

async def request_package(client: httpx.AsyncClient, name: str, range: str) -> tuple[VersionedPackage, dict]:
    url = f"{NPM_REGISTRY_URL}/{name}"
    response = await client.get(url)

    try:
        response.raise_for_status()
        npm_package = response.json()
    
    except httpx.HTTPStatusError as e:
        raise  # TODO: Handle different errors for different responses

    versions = list(npm_package["versions"].keys())
    version = semver.min_satisfying(versions, range)
    version_record = npm_package["versions"][version]

    return VersionedPackage(
        name=version_record["name"],
        version=version_record["version"],
        description=version_record["description"],
    ), version_record.get("dependencies", {})

async def fetch_all_dependencies(dependencies: dict) -> list[VersionedPackage]:
    tasks = [
        get_package(name=dep_name, range=dep_range)
        for dep_name, dep_range in dependencies.items()
    ]
    return await asyncio.gather(*tasks)
```

File: tests/test_package.py
Comment (MichaelYochpaz) [inline tests/test_package.py:1]: Please add additional unit-tests to cover some edge-cases, such as an invalid package name being used, a package with no dependencies, and a package with recursive dependencies (one of its dependencies uses it as a dependency).

Also, please move expected JSON results to a separate JSON file (which is loaded using `open`) to keep the tests file clean and organized.

File: packages/modules/npm.py
Comment (MichaelYochpaz) [reply packages/modules/npm.py:1]: If a package's dependency uses the package as one of its dependencies (recursive dependencies), this code will be stuck in an endless loop.

This should be resolved by caching, but make sure you add a robust solution to avoid this issue (even if, for example, the cache size runs out)
